contracts/zaps/ZapYearnShares.sol

```diff
@@ -0,0 +1,104 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.0;
+
+import "../interfaces/IYearnVaultV2.sol";
+import "../libraries/Authorizable.sol";
+import "../interfaces/IERC20.sol";
+import "../interfaces/ITranche.sol";
+import "hardhat/console.sol";
+
+contract ZapYearnShares is Authorizable {
+    // Store the accessibility state of the contract
+    bool public isFrozen = false;
+    // Tranche factory address for Tranche contract address derivation
+    address internal immutable _trancheFactory;
+    // Tranche bytecode hash for Tranche contract address derivation.
+    // This is constant as long as Tranche does not implement non-constant constructor arguments.
+    bytes32 internal immutable _trancheBytecodeHash;
+
+    /// @param __trancheFactory Address of the TrancheFactory contract
+    /// @param __trancheBytecodeHash Hash of the Tranche bytecode.
+    constructor(address __trancheFactory, bytes32 __trancheBytecodeHash)
+        Authorizable()
+    {
+        _authorize(msg.sender);
+        _trancheFactory = __trancheFactory;
+        _trancheBytecodeHash = __trancheBytecodeHash;
+    }
+
+    /// @dev Requires that the contract is not frozen
+    modifier notFrozen() {
+        require(!isFrozen, "Contract frozen");
+        _;
+    }
+
+    /// @dev Allows an authorized address to freeze or unfreeze this contract
+    /// @param _newState True for frozen and false for unfrozen
+    function setIsFrozen(bool _newState) external onlyAuthorized() {
+        isFrozen = _newState;
+    }
+
+    /// @notice Mints a Principal/Interest token pair from yearn vault shares.
+    ///      then returns the tokens to the caller.
+    /// @param _underlying The underlying ERC20 token contract of the yearn vault.
+    /// @param _vault The address of the target yearn vault.
+    /// @param _amount The amount of yearn shares to turn into tokens
+    /// @param _expiration The expiration time of the Tranche contract.
+    /// @param _position The contract which manages pooled deposits.
+    /// @param _maxLoss The maximum acceptable loss to sustain on withdrawal.
+    /// @return returns the minted amounts of principal and yield tokens (PT and YT)
+    function zapSharesIn(
+        IERC20 _underlying,
+        IYearnVault _vault,
+        uint256 _amount,
+        uint256 _expiration,
+        address _position,
+        uint256 _maxLoss
```

Comment:
on the fence about having this be settable here 🤔

Reply:
I think that if the yearn vault has loss we shouldn't be allowing tranche deposits.

Reply:
I'm good with defaulting to 0